api/api_public_transaction_pool.go

```diff
@@ -552,3 +552,46 @@ func resend(s *PublicTransactionPoolAPI, ctx context.Context, sendArgs NewTxArgs
 
 	return common.Hash{}, fmt.Errorf("Transaction %#x not found", matchTx.Hash())
 }
+
+// RecoverTransaction recovers the sender address from a signed raw transaction.
+// The signature is validated against the sender account's key configuration at the given block number.
+func (s *PublicTransactionPoolAPI) RecoverTransaction(ctx context.Context, encodedTx hexutil.Bytes, blockNumber rpc.BlockNumber) (common.Address, error) {
+	tx := new(types.Transaction)
+	if err := rlp.DecodeBytes(encodedTx, tx); err != nil {
+		return common.Address{}, err
+	}
+
+	signer := types.MakeSigner(s.b.ChainConfig(), s.b.CurrentBlock().Number())
+	state, _, err := s.b.StateAndHeaderByNumber(ctx, blockNumber)
+	if err != nil {
+		return common.Address{}, err
+	}
+	_, err = tx.ValidateSender(signer, state, blockNumber.Uint64())
+	if err != nil {
+		return common.Address{}, err
+	}
+	return signer.Sender(tx)
+}
+
+// RecoverMessage validates that the message is signed by one of the keys in the given account.
+// Returns the recovered signer address, which may be different from the account address.
+func (s *PublicTransactionPoolAPI) RecoverMessage(
+	ctx context.Context, address common.Address, data, sig hexutil.Bytes, blockNumber rpc.BlockNumber,
+) (common.Address, error) {
+	pubkey, err := klayEthEcRecover(data, sig)
+	if err != nil {
+		return common.Address{}, err
+	}
+
+	state, _, err := s.b.StateAndHeaderByNumber(ctx, blockNumber)
+	if err != nil {
+		return common.Address{}, err
+	}
+	key := state.GetKey(address)
+	// If the key is the AccountKeyTypeLegacy, confirmation will be delegated to the user
+	// by returning the recovered address
+	if key.ValidateMember(pubkey, address) {
+		return crypto.PubkeyToAddress(*pubkey), nil
```

Comment:
I think option 1 is better. I think `ecrecover`-ed address doesn't imply anything.

Reply:
What you mean option 1?

Reply:
in the pr description

```
need to discuss what to return:
option1 - the address
option2 - extracted public key from the signature to avoid confusion with the account address
```

Reply:
option 1 refers to 'the recovered address' and option 2 refers to the 'recovered public key'. Both are from the signature. Me and @kjeom had a discussion about this API, and we concluded that returning the recovered address/pubkey is meaningful in some context - where a webapp has to collect signatures from distinct individuals of their shared account, so the webapp has to know which key signed the message.

Reply:
I see. Please clarify the difference of the address parameter and the returned address in the API docs.